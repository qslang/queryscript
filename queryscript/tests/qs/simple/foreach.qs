import users from schema;

SELECT for item in [id, org_id] {
	item
} FROM users;

SELECT for item in [id, org_id] {
	item AS item
} FROM users;


SELECT for item in [id, org_id] {
    users.*
} FROM users;

SELECT for (item in [id, org_id]) {
    item,
    item+1
} FROM users;

SELECT for item in [1, 2] {
    item
} FROM users;

let slices = ['month', 'day'];

select
    for item in slices {
        item
    }
;

SELECT for item in ['id', 'org_id'] {
    f"{item}" AS f"metric_{item}"
} FROM users;

let strings = ['id', 'org_id'];
let numbers = ['id', 'org_id'];

strings;
SELECT strings;

numbers;
SELECT numbers;

SELECT for item in numbers {
    item
} FROM users;


SELECT for item in strings {
    item
} FROM users;

-- Idents
SELECT for item in strings {
    f"{item}"
} FROM users;

let id_types = ['', 'org_'];
SELECT for item in id_types {
    f"{item}id"
} FROM users;

SELECT for item in [id, org_id] {
    item
} FROM users
GROUP BY for item in [id, org_id] {
    item
};

select
    case
        for item in strings {
            when f"{item}"=1 then item
        }
    end as "foo"
from users
;

-- Functions

fn loop_fixed(fields [text]) {
    SELECT
        for f in fields {
            f"{f}"
        }
    FROM users
}
loop_fixed(['id']);

fn loop_generic<R>(fields [text], t R) {
    SELECT
        for f in fields {
            f"{f}"
        }
    FROM t
}
loop_generic(['id', 'org_id'], users);
