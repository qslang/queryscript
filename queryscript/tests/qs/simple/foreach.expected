{
    "compile_errors": [],
    "decls": {
        "let id_types": [Utf8],
        "let loop_fixed": λ {
        	fields [Utf8],
        } -> [{
        	id Int32,
        }],
        "let loop_generic": ∀ "__Return", "R" λ {
        	fields [Utf8],
        	t "R",
        } -> "__Return",
        "let numbers": [Utf8],
        "let slices": [Utf8],
        "let strings": [Utf8],
        "let users": [{
        	id Int32,
        	org_id Int32,
        	name Utf8,
        	active Boolean,
        }],
    },
    "queries": [
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "org_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | org_id |\n|----|--------|\n| 1  | 1      |\n| 2  | 1      |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "item",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "item",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| item | item |\n|------|------|\n| 1    | 1    |\n| 2    | 1    |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "org_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "name",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "active",
                                type_: Atom(
                                    Boolean,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "org_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "name",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "active",
                                type_: Atom(
                                    Boolean,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | org_id | name | active | id | org_id | name | active |\n|----|--------|------|--------|----|--------|------|--------|\n| 1  | 1      | Foo  | true   | 1  | 1      | Foo  | true   |\n| 2  | 1      | Bar  | false  | 2  | 1      | Bar  | false  |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "\"users\".\"id\" + 1",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "org_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "\"users\".\"org_id\" + 1",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | \"users\".\"id\" + 1 | org_id | \"users\".\"org_id\" + 1 |\n|----|------------------|--------|----------------------|\n| 1  | 2                | 1      | 2                    |\n| 2  | 3                | 1      | 2                    |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "CAST(1 AS BIGINT)",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "CAST(2 AS BIGINT)",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| CAST(1 AS BIGINT) | CAST(2 AS BIGINT) |\n|-------------------|-------------------|\n| 1                 | 2                 |\n| 1                 | 2                 |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "'month'",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "'day'",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| 'month' | 'day' |\n|---------|-------|\n| month   | day   |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "metric_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "metric_org_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| metric_id | metric_org_id |\n|-----------|---------------|\n| 1         | 1             |\n| 2         | 1             |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Atom(
                        Utf8,
                    ),
                ),
                value: "[id, org_id]",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "strings",
                                type_: List(
                                    Atom(
                                        Utf8,
                                    ),
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| strings      |\n|--------------|\n| [id, org_id] |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Atom(
                        Utf8,
                    ),
                ),
                value: "[id, org_id]",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "numbers",
                                type_: List(
                                    Atom(
                                        Utf8,
                                    ),
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| numbers      |\n|--------------|\n| [id, org_id] |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "'id'",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "'org_id'",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| 'id' | 'org_id' |\n|------|----------|\n| id   | org_id   |\n| id   | org_id   |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "'id'",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "'org_id'",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| 'id' | 'org_id' |\n|------|----------|\n| id   | org_id   |\n| id   | org_id   |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "org_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | org_id |\n|----|--------|\n| 1  | 1      |\n| 2  | 1      |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "org_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | org_id |\n|----|--------|\n| 1  | 1      |\n| 2  | 1      |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "org_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | org_id |\n|----|--------|\n| 1  | 1      |\n| 2  | 1      |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "foo",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| foo    |\n|--------|\n| id     |\n| org_id |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id |\n|----|\n| 1  |\n| 2  |",
                viz: None,
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "org_id",
                                type_: Atom(
                                    Int32,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | org_id |\n|----|--------|\n| 1  | 1      |\n| 2  | 1      |",
                viz: None,
            },
        ),
    ],
}